---
title: 'Week 9: DESeq2 Analysis'
author: "Chantal Rabay"
date: "2/27/2022"
output: pdf_document
---

Week 9 consists of analysis of RNASeq data using two input files: a counts matrix and associated metadata

# 1. Bioconductor and DESeq2 setup
Used the following code in the R console in order to install Bioconductor packages


# 2. Import countData and colData

```{r}
#Import these files
counts <- read.csv("airway_scaledcounts.csv", row.names = 1)
metadata <- read.csv("airway_metadata.csv")
```

Inspecting the data
```{r}
head(counts)
```

```{r}
head(metadata)
```

### [Q1] How many genes are in this dataset?
There are 38,694 genes in this dataset.

### [Q2] How many ‘control’ cell lines do we have?
There are 4 'control' cell lines.

# 3. Toy differential gene expression
Exploratory differential analysis for demonstration. Never do differential analysis this way.

```{r}
#Getting just the counts for non-drug treated samples first
control <- metadata[metadata[,"dex"]=="control",]
control.counts <- counts[ ,control$id]
control.mean <- rowSums( control.counts )/4 
head(control.mean)
```

```{r}
#Same thing as above but using dplyr
library(dplyr)
control <- metadata %>% filter(dex=="control")
control.counts <- counts %>% select(control$id) 
control.mean <- rowSums(control.counts)/4
head(control.mean)
```

### [Q3] How would you make the above code in either approach more robust?

### [Q4] Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

```{r}
treated.mean <- rowMeans(counts[, metadata[metadata$dex == "treated", ]$id])
```

Create a data frame containing both of the mean counts 
```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

Use colSums() to get the sums of mean counts 
```{r}
colSums(meancounts)
```

### [Q5a] Create a scatter plot showing the mean of the treated samples against the mean of the control samples. 

```{r}
plot(meancounts[, 1:2])
```
### [Q5b] You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot?
The geom_point() function.
```{r}
library(ggplot2)

ggplot(meancounts) +
  aes(control.mean, treated.mean) +
  geom_point()
```
### [Q6] Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

```{r}
plot(meancounts[, 1:2], log = "xy")
```



```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```
Filtering data to remove genes with NaN values or -Inf values

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```
### [Q7] What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?
The arr.ind argument allows us to return the row and column positions where there are TRUE values, which stand for rows or columns with zero counts. The unique() function removes any repeated elements which ensures that rows are not counted twice.

A common rule of thumb in this field is to first focus on large changes with a log2fc cutoff of +2 or -2.

```{r}
up.ind <- sum(mycounts$log2fc > 2)
down.ind <- sum(mycounts$log2fc < (-2))
print(up.ind)
print(down.ind)
```


### [Q8] Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level? 
250 up regulated genes

### [Q9] Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level? 
367 down regulated genes

### [Q10] Do you trust these results? Why or why not?
No, I do not trust these results because when looking at the fold change it can be large but not statistically significant. Based on p-values we need to determine which differences are statistically significant.

# 4. DESeq2 analysis 
```{r}
library("DESeq2")
citation("DESeq2")
```

## Importing Data

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

## DESeq analysis

```{r}
#Running results before the analysis creates an error.
#results(dds)
```

```{r}
#Getting help on the DESeq function
?DESeq
```

```{r}
dds <- DESeq(dds)
```

## Getting Results
```{r}
res <- results(dds)
res
```

Converted the res object to a data frame using as.data.frame() function and view() function in the console.

Summarizing basic tallies using the summary function.
```{r}
summary(res)
```

creating a result with a p-value of 0.5
```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

# 5. Adding Annotation Data

Download and install Bioconductor's main annotation packages to help with mapping: organism annotation package ("org") for Homo sapiens (“Hs”), organized as an AnnotationDbi database package (“db”), using Entrez Gene IDs (“eg”).
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```

```{r}
#pulling up documentation with help() command
help("REFSEQ")
```

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```
```{r}
head(res)
```

### [Q11] Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res$genename.
```{r}
#Adding the Entrez ID
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")
```
```{r}
#Adding the UniProt Accession
res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")
```
```{r}
#Adding the GENENAME
res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")
```
```{r}
head(res)
```

Arrange and view results by adjusted p-value
```{r}
ord <- order( res$padj )
#View(res[ord,])
head(res[ord,])
```

Writing out the ordered significant results with annotations.
```{r}
write.csv(res[ord,], "deseq_results.csv")
```

# 6. Data Visualization 
## Volcano Plots 
```{r}
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

Using EnhancedVolcano Bioconductor package for more customization.
```{r}
library(EnhancedVolcano)
```

```{r}
x <- as.data.frame(res)

EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```

# 7. Pathway Analysis
## Pathway Analysis With R and Bioconductor

Installed pathview, gage, and gageData in R console.

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```
Running gage pathway analysis
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```
```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa05310", kegg.native=FALSE)
```

### [Q12] Can you do the same procedure as above to plot the pathview figures for the top 2 down-reguled pathways?
```{r}
#Pathview figure for top 1 down regulated pathway
pathview(gene.data=foldchanges, pathway.id="hsa05332", kegg.native=FALSE)
```

```{r}
#Pathview figure for top 1 down regulated pathway
pathview(gene.data=foldchanges, pathway.id="hsa04940", kegg.native=FALSE)
```

## OPTIONAL: Plotting Counts for Genes of Interest
```{r}
#Getting the gene id for CRISPLD2 gene
i <- grep("CRISPLD2", res$symbol)
res[i,]
```

```{r}
rownames(res[i,])
```

```{r}
#plotting counts
plotCounts(dds, gene="ENSG00000103196", intgroup="dex")
```
```{r}
# Return the data
d <- plotCounts(dds, gene="ENSG00000103196", intgroup="dex", returnData=TRUE)
head(d)
```
Plot a boxplot
```{r}
boxplot(count ~ dex , data=d)
```
Plot in ggplot2
```{r}
library(ggplot2)
ggplot(d, aes(dex, count, fill=dex)) + 
  geom_boxplot() + 
  scale_y_log10() + 
  ggtitle("CRISPLD2")
```










