---
title: 'Class 12: RNA-Seq Analysis Mini-Project'
author: "Chantal Rabay"
date: "2/28/2022"
output: pdf_document
---

# Section 1. Differential Expression Analysis 
```{r}
library(DESeq2)
```

Load our data files.
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```
```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

### \textcolor{red}{Q. Complete the code below to remove the troublesome first column from countData.}
```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```

### \textcolor{red}{Q. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns).}

```{r}
sumrow <- as.data.frame(rowSums(countData))
zerosum <- which(sumrow[,1] == 0, arr.ind=TRUE)

# Filter count data where you have 0 read count across all samples.
countData = countData[-zerosum, ]
head(countData)
```

## Running DESq2
```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
dds
```
```{r}
res = results(dds)
```

### \textcolor{red}{Q. Call the summary() function on your results to get a sense of how many genes are up or down-regulated at the default 0.1 p-value cutoff.}

```{r}
summary(res)
```

## Volcano Plot

```{r}
plot( res$log2FoldChange, -log(res$padj) )
```
### \textcolor{red}{Q. Improve this plot by completing the below code, which adds color and axis labels}

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.1) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```

## Adding Gene Annotation

### \textcolor{red}{Q. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below.}

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```
```{r}
columns(org.Hs.eg.db)
```

Adding Symbol annotation
```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")
```
Adding EntrezID annotation
```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")
```

Adding Gene name annotation
```{r}
res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")
```
```{r}
head(res, 10)
```

### \textcolor{red}{Q.Finally for this section let's reorder these results by adjusted p-value and save them to a CSV file in your current project directory.}

```{r}
res = res[order(res$pvalue),]
write.csv(res, file ="deseq_results.csv")
```

# Section 2. Pathway Analysis 

## KEGG pathways 
```{r}
library(pathview)
```

Loading packages and setting up the KEGG data sets we need.
```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```
Fold change results from the DESeq2 analysis are stored in res$log2FoldChange
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Running the gage pathways analysis
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres)
```

Look at the first few down(less) pathway results:
```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

Use pathway to make a pathway plot with our RNA Seq expression results
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

# Section 3. Gene Ontology (GO)
Similar process with Gene Ontology 
```{r}
data(go.sets.hs)
data(go.subs.hs)
```

```{r}
# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

# Section 4. Reactome Analysis
Over-representation enrichment analysis and pathway-topology analysis with Reactome
```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```


### \textcolor{red}{Q. What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?}

The pathway with the most significant p-value is the endosomal/vacuolar pathway. The results are not the same. This may just be due to differences in the databases and the data and methods that they use. The second most significant pathway in reactome was cell cycle, which was the most significant pathways listed from KEGG.
